feat: add dashboard option `allowAnonymousUser`

dblythy opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Parse Dashboard assumes security if it detects a remote server, which is unreliable. This adds the option `allowAnonymousUser` which defaults to false.

Related issue: #2065

### Approach
<!-- Add a description of the approach in this PR. -->

- Adds a Parse Dashboard option `allowAnonymousUser` that defaults to false. Setting this to true allows Dashboard to run with no users.
- Adds ability to specify options via command line or via config file
- Adds console.error to frontend so solutions to errors can be identified by developers

This also add additional security for when dashboard is run on localhost with no options. Previously Dashboard was effectively assuming the `dev` parameter if localhost. This PR will only run Dashboard in `dev` mode if it's implicitly set. To get previous localhost behaviour, the `dev` parameter is required to be set.

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete TODOs that do not apply to this PR.
-->

- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

mtrezza committed:
ci: bump environment

semantic-release-bot committed:
chore(release): 4.0.0-alpha.16 [skip ci]

* adding internal class (e.g. `_User`) fails due to prefixed underscore ([#2036](https://github.com/ParsePlatform/parse-dashboard/issues/2036)) ([f80bd07](https://github.com/ParsePlatform/parse-dashboard/commit/f80bd07a42b19fc4fa2632e0147fa72812a87c2f))

snyk-bot committed:
refactor: upgrade typescript from 4.5.4 to 4.5.5 (#2038)

snyk-bot committed:
refactor: upgrade otpauth from 7.0.9 to 7.0.10 (#2044)

snyk-bot committed:
refactor: upgrade graphql from 16.2.0 to 16.3.0 (#2043)

semantic-release-bot committed:
chore(release): 4.0.0-alpha.17 [skip ci]

# [4.0.0-alpha.17](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.16...4.0.0-alpha.17) (2022-02-23)

### Bug Fixes

* security upgrade prismjs from 1.26.0 to 1.27.0 ([#2047](https://github.com/ParsePlatform/parse-dashboard/issues/2047)) ([3afb24e](https://github.com/ParsePlatform/parse-dashboard/commit/3afb24e708a69560732a725574953333431c1ca9))

snyk-bot committed:
refactor: upgrade @babel/runtime from 7.16.7 to 7.17.0 (#2048)

snyk-bot committed:
fix: upgrade @babel/runtime from 7.17.0 to 7.17.2 (#2055)

Snyk has created this PR to upgrade @babel/runtime from 7.17.0 to 7.17.2.

See this package in npm:
https://www.npmjs.com/package/@babel/runtime

See this project in Snyk:
https://app.snyk.io/org/acinader/project/3e039b91-2450-4b56-8420-baf56cab388e?utm_source=github&utm_medium=referral&page=upgrade-pr

semantic-release-bot committed:
chore(release): 4.0.0-alpha.18 [skip ci]

# [4.0.0-alpha.18](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.17...4.0.0-alpha.18) (2022-03-02)

### Bug Fixes

* upgrade @babel/runtime from 7.17.0 to 7.17.2 ([#2055](https://github.com/ParsePlatform/parse-dashboard/issues/2055)) ([93335e9](https://github.com/ParsePlatform/parse-dashboard/commit/93335e9ab52e7b1f203b7939abde5c12b42dd2a7))

snyk-bot committed:
refactor: upgrade body-parser from 1.19.1 to 1.19.2 (#2057)

snyk-bot committed:
fix: upgrade express from 4.17.2 to 4.17.3 (#2058)

Snyk has created this PR to upgrade express from 4.17.2 to 4.17.3.

See this package in npm:
https://www.npmjs.com/package/express

See this project in Snyk:
https://app.snyk.io/org/acinader/project/3e039b91-2450-4b56-8420-baf56cab388e?utm_source=github&utm_medium=referral&page=upgrade-pr

Co-authored-by: Manuel <5673677+mtrezza@users.noreply.github.com>

semantic-release-bot committed:
chore(release): 4.0.0-alpha.19 [skip ci]

# [4.0.0-alpha.19](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.18...4.0.0-alpha.19) (2022-03-10)

### Bug Fixes

* upgrade express from 4.17.2 to 4.17.3 ([#2058](https://github.com/ParsePlatform/parse-dashboard/issues/2058)) ([f8dc602](https://github.com/ParsePlatform/parse-dashboard/commit/f8dc6027f57143d72083f593455b23e304de934a))

semantic-release-bot committed:
chore(release): 4.0.0-alpha.20 [skip ci]

# [4.0.0-alpha.20](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.19...4.0.0-alpha.20) (2022-03-16)

### Features

* change string filter description ([#2059](https://github.com/ParsePlatform/parse-dashboard/issues/2059)) ([bb1e184](https://github.com/ParsePlatform/parse-dashboard/commit/bb1e184272497749630964b4bd9e610d2d4cd328))

snyk-bot committed:
fix: upgrade otpauth from 7.0.10 to 7.0.11 (#2061)

Snyk has created this PR to upgrade otpauth from 7.0.10 to 7.0.11.

See this package in npm:
https://www.npmjs.com/package/otpauth

See this project in Snyk:
https://app.snyk.io/org/acinader/project/3e039b91-2450-4b56-8420-baf56cab388e?utm_source=github&utm_medium=referral&page=upgrade-pr

semantic-release-bot committed:
chore(release): 4.0.0-alpha.21 [skip ci]

# [4.0.0-alpha.21](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.20...4.0.0-alpha.21) (2022-03-18)

### Bug Fixes

* upgrade otpauth from 7.0.10 to 7.0.11 ([#2061](https://github.com/ParsePlatform/parse-dashboard/issues/2061)) ([c379306](https://github.com/ParsePlatform/parse-dashboard/commit/c3793061a186d49ab09e8bd448e167078d1e4df7))

dblythy committed:
feat: add dashboard option `allowAnonymousUser`

dblythy committed:
Update app.js

dblythy committed:
Update Dashboard.js

mtrezza commented on Parse-Dashboard/app.js:
```diff
@@ -87,21 +87,29 @@ module.exports = function(config, options) {
-      if (!options.dev && !requestIsLocal) {
-        if (!req.secure && !options.allowInsecureHTTP) {
+      for (const key in options) {
+        if (options[key] && !config[key]) {
```
Does this work if an option has value `false`?

dblythy replied:
Good thinking

mtrezza commented on Parse-Dashboard/app.js:
```diff
@@ -87,21 +87,29 @@ module.exports = function(config, options) {
+            success: false,
+            error:
+              "Parse Dashboard can only be remotely accessed via HTTPS.",
+            log: "Parse Dashboard can only be remotely accessed via HTTPS. If you are running locally, use the --dev parameter to bypass allowInsecureHTTP.",
```
I think "bypass" is ambiguous, it means "undefined state" in this case because it doesn't indicate whether the `allowInsecureHTTP` is true or false when it's "bypassed".

That wording is also used elsewhere in the docs. Maybe we can be more clear and say it "overrides the option to `false`" or something like that?

RaschidJFR committed:
ci: refactor lint CI task and fix lint issues (#2004)

dblythy committed:
review feedback

mtrezza commented:
Still some checks failing. Please request another review once this is ready.

dblythy committed:
Update app.js

dblythy committed:
Update dashboard.e2e.test.js

mtrezza commented:
Is this a breaking change? Does an existing deployment need a reconfiguration to maintain the same security?

dblythy commented:
I would say so, an existing localhost configuration would need to be changed but remote servers should function the same as before

mtrezza commented:
I guess our deprecation policy should be (or is in practice already) extending to Parse Dashboard. So breaking changes should probably happen accumulated once a year only as well. Otherwise we may have a breaking change every other month.

Is there a way to make this non-breaking? For example, if the new option is not set, fall back to the old behavior.

dblythy commented:
The only way I think is if we set `dev` to true by default, but that brings security implications

snyk-bot committed:
refactor: upgrade inquirer from 8.2.0 to 8.2.1 (#2079)

mtrezza commented:
Yes, I think we should avoid that

mtrezza commented:
Could you rebase this on alpha?

semantic-release-bot committed:
chore(release): 4.1.1-alpha.1 [skip ci]

## [4.1.1-alpha.1](https://github.com/ParsePlatform/parse-dashboard/compare/4.1.0...4.1.1-alpha.1) (2022-04-04)

### Bug Fixes

* security upgrade js-beautify from 1.14.0 to 1.14.1 ([#2077](https://github.com/ParsePlatform/parse-dashboard/issues/2077)) ([e4ea787](https://github.com/ParsePlatform/parse-dashboard/commit/e4ea7879d88173b02d66b1339ba98805255ba82c))
* security vulnerability bump minimist from 1.2.5 to 1.2.6 ([#2070](https://github.com/ParsePlatform/parse-dashboard/issues/2070)) ([3d0407e](https://github.com/ParsePlatform/parse-dashboard/commit/3d0407ebd75051bbbe6f0a2aba87b26475e901b9))

dblythy committed:
alpha

dblythy committed:
revert changelog

snyk-bot committed:
refactor: upgrade commander from 9.0.0 to 9.1.0 (#2089)

snyk-bot committed:
refactor: upgrade @babel/runtime from 7.17.2 to 7.17.7 (#2084)

snyk-bot committed:
refactor: upgrade graphiql from 1.6.0 to 1.7.1 (#2083)

snyk-bot committed:
refactor: upgrade @babel/runtime from 7.17.7 to 7.17.8 (#2090)

mtrezza committed:
Merge branch 'alpha' into allowAnonymousUser

mtrezza reviewed (commented):
It seems the discussion got stuck at how to avoid making this a breaking change. Once this has been figured out please request another review.
